### ISSUE TITLE
Discriminator returns nans during training.

### ISSUE THREAD

**Issue by vasl12** [neel-dey/Atlas-GAN#7]
Hi! 
I am trying to run the repo on a custom dataset to generate conditional age and csf_volume atlases. 
I am encountering 2 different issues. One is that at random points, the discriminator is returning nan values, and as a result, the training fails. I tried different combinations of hyperparameters, but I cannot pinpoint where this behaviour comes from. 
Secondly, I tried to continue the training from a previous checkpoint and at some circumstances the training progresses until it reaches another nan value. 
Using this strategy I managed to train the network for over 100 epochs, but when I am trying to use this for inference, it seems that the generator generates the same image no matter what the condition is (mode collapse?).
I would therefore like to ask you if you have encountered this in the past or whether you have any ideas on how to change the hyperparams to enable learning.

Thanks a lot for the help.

**Comment by neel-dey** [comment]
Hi @vasl12,

Thanks for your interest! Happy to help.

Some questions to debug:

(1) I actually saw your Diff-Def paper through gscholar citation alerts a few months ago (congrats! it's very nice work). 

If I'm reading Table 3 correctly, your implementation of our method seems to be using the MSE registration loss and very different regularization weights. Our method uses the LNCC registration loss and would require a completely different set of hyperparameters and retuning to weigh the various regularizers appropriately. For example, with MSE for registration and lambda_Reg=0.001, the GAN loss would be significantly overweighted, leading to unstable training.

Can you try running the method in this repo as is with the default hyperparameters and loss?

(2) If that doesn't resolve it, can you try setting `losswt_gan` and `losswt_gp` to 0? That'll turn off GAN training and produces good conditional templates as well (primarily due to the [different template synthesis architecture](https://github.com/neel-dey/Atlas-GAN/blob/main/src/networks.py#L244)). 

If training that for 150K--200K iterations is stable and produces conditionally varying templates, I would guess that it's primarily because of (1). If it's still unstable, I would suspect a bad data batch or something, especially given that you report that resuming training changes behavior.

Happy to discuss further :)

**Comment by vasl12** [comment]
Hi @neel-dey, 

Thanks for you quick reply and your kind words. :)

In response to (1) we are using the hyperparameters you suggest in the paper and in the repo. We have even tried the different sets you suggest for each of the 2 datasets you are using, but the problem is still present.

Thank you for the suggestions in point (2) we will give them a go!!

I can provide you with some extra information that might be of use. :)

We use batch size of 1 and we double checked that we don't have inappropriate values for both images and conditions. 

What we see is that somehow the training results in learning the average shape of the average of the conditions of the dataset. In other words, when we restrict the range of the condition in the upper and lower part (e.g. 0.0-0.5 and 0.5-1.0) it seems that the network is partially learning larger or smaller ventricles, but when we are using the whole dataset it cannot learn the conditioning effectively resulting in always the same brain shape. 
We also double checked that we are using more or less similar amount of data for each condition. 

Finally, we noticed you choose 1000 steps per epoch. Is there any intuition on why to use so many steps?

Thanks a lot for the help and the intuitions!

**Comment by neel-dey** [comment]
Ah sorry, replies will be a bit slow this week (CVPR deadline, ICLR rebuttal, NeurIPS camera-ready video, etc.)

> In response to (1) we are using the hyperparameters you suggest in the paper and in the repo. We have even tried the different sets you suggest for each of the 2 datasets you are using, but the problem is still present.

Alright, so to confirm: you're using the repo as is and all the defaults and only changing the input data?

In that case, this is new as previous user-reported issues were related to training instability or artifacts, I haven't had users report NaNs or encountered them myself during training before. Some standard debugging questions:

1. Are your images intensity normalized between 0 and 1?
2. Are your continuous input conditions (e.g. age) normalized between 0 and 1?
3. Are your categorical input conditions (e.g. disease) represented as one-hot vectors?
4. Are your images rigid/affine pre-aligned?
5. Can you also please post the tensorboard loss curves?
6. Does the NaN specifically occur first in the discriminator loss or does it break first in the NCC loss (subsequently breaking everything else)?
7. Is it possible that some input failed rigid/affine alignment? Sometimes they go all-black/off-grid if the step size is too high. If so, are these the samples that cause the NaN?

I think we'll have more clarity after the experiment below:

> Thank you for the suggestions in point (2) we will give them a go!!

Please let me know when this finishes and if it worked for your setup. That way we can pinpoint whether this is related to the GAN loss or something with how the data is encoded.

> What we see is that somehow the training results in learning the average shape of the average of the conditions of the dataset. In other words, when we restrict the range of the condition in the upper and lower part (e.g. 0.0-0.5 and 0.5-1.0) it seems that the network is partially learning larger or smaller ventricles, but when we are using the whole dataset it cannot learn the conditioning effectively resulting in always the same brain shape.

- Please see questions 2 & 3 above. Also, what are the conditions here? It's plausible that something in how they're encoded needs to be different from what we did in the paper.
- Generally, it starts off with the average shape and then starts conditionally varying within 5-10K iterations if I remember correctly, so this is probably related to the issues above. Once we figure out what's the bug this should go away.

> We also double checked that we are using more or less similar amount of data for each condition.

Can you please try training with and without the `--oversample` flag? 

> Finally, we noticed you choose 1000 steps per epoch. Is there any intuition on why to use so many steps?

Yes, I found that high-quality atlases require about 150--200K iterations with batchsize 1 to converge (with both this method and Adrian's 2019 paper). So epochs x steps = 150 x 1000 here

Let's hope we find something :)

**Comment by neel-dey** [comment]
Hi @vasl12, did you get a chance to look into this?

**Comment by vasl12** [comment]
Hi @neel-dey, 
First of all, thanks for your help and the suggestions. We had already preprocessed the data appropriately, so unfortunately this was not the issue. 

However, we managed to pinpoint where the issue was coming from. Although the brains were registered, some of them have shapes that are quite different than the others (either due to their anatomy or due to skull stripping failure). As a result the more "weird" shapes the training was encountering the more it failed.

To account for this we developed a filtering process where we measure how different each of the brain is to an MNI atlas and by defining a threshold we managed to keep only the most similar brains and the process is running without an issue now.

Thank you very much one more time for the help and the suggestions!

**Comment by neel-dey** [comment]
Hm, that's odd, as we had variably shaped inputs as well and didn't encounter NaNs in training. 

For example, dHCP had variable appearances and shapes (see below) and Predict-HD also had significant variability (age range of ~20--90) and had a few skullstripping failures as well. In case you haven't seen it before, you might be interested in [synthstrip](https://surfer.nmr.mgh.harvard.edu/docs/synthstrip/), it is generally much better at skullstripping than BET/ROBEX/etc. and might be helpful to clean up any outliers.


<img src="https://github.com/user-attachments/assets/5c905712-9a69-481f-93cd-933540ba07d6" width="480">


Either way, great that it's training now! Please feel free to reopen the issue or start a new one if you run into any other issues :)
